### ISSUE TITLE
PodSecurityContext API compatibility broken between 1.6 and 1.7

### ISSUE THREAD

**Issue by kow3ns** [kubernetes/kubernetes#47448]
** BUG REPORT** 

**Kubernetes version**: 
1.7.x with a 1.6.x kubectl or any version of helm based on a vendored client at that version.

**Environment**: 
This is an API issue affecting all environments.

**What happened**:
When you attempt to apply or create a manifest containing any containers (e.g. DaemonSet, StatefulSet, Deployment) with a PodTemplateSpec with any Container having PodSecurityContext that attempts to set RunAsUser or FSGroup you will get an error message indicating that those fields are invalid.

**What you expected to happen**:
I expect to be able to create a Pods that do not run as root.

**How to reproduce it**:
Build and install a 1.6.3 client and kube-up a cluster off of master. Attempt to create a manifest that uses the described features of security context. [Here](https://github.com/kow3ns/kubernetes-zookeeper/blob/master/manifests/zookeeper.yaml) is an example.

**Anything else we need to know**:
[This](https://github.com/kubernetes/kubernetes/commit/9440a68744ea97a45ceeef6d75466405ac101ea1) looks like the commit that modifies the v1/types.go in an incompatible way.

**Comment by kow3ns** [comment]
@kubernetes/api-approvers 
@kubernetes/api-reviewers

**Comment by kow3ns** [comment]
@kubernetes/sig-api-machinery-api-reviews 
@kubernetes/sig-api-machinery-bugs

**Comment by liggitt** [comment]
is this only with kubectl and client-side validation? if you specify `--validate=false`, can it persist correctly? As far as I can tell, that PR is wire-compatible with 1.6

**Comment by kow3ns** [comment]
@liggit yes its wire compatible but validation fails.

**Comment by liggitt** [comment]
see https://github.com/kubernetes/kubernetes/issues/39434 for known issues with client-side validation

**Comment by liggitt** [comment]
I'm not sure what to do with this one... bleeding the go field types out into openapi/swagger specs in https://github.com/kubernetes/kubernetes/commit/9440a68744ea97a45ceeef6d75466405ac101ea1 shouldn't have been necessary and was not a great idea, but I also don't know whether our API compatibility only covers the actual API or also the generated openapi doc.

**Comment by kow3ns** [comment]
@liggitt  that is fair. I'm not sure either :). If you want to close and differ to #39434 that is fine. I'm trying to figure a way to work around this for helm charts now. I'm not sure if helm allows you to disable client side validation. If not, my concern is that this will break any helm chart that wants to set a user that is non-root for the 1.7 release. Even the next release, 2.5.x, will still be vendoring the 1.6.x client.

**Comment by kow3ns** [comment]
@foxish @erictune thoughts?

**Comment by liggitt** [comment]
cc @kubernetes/sig-cli-bugs on client-side validation issues. aside from this, is there anything we should do now to make the 1.7 CLI tolerate unknown custom types in validation.

**Comment by ericchiang** [comment]
We've hit the same issue #47560 when trying to update bootkube to 1.7.

TL;DR you can't create the following deployment with kubectl 1.6 against a 1.7 cluster.

```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: nginx-deployment
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.7.9
      securityContext:
        runAsNonRoot: true
        runAsUser: 65534 # Fails validation
```

Error message for a similar deployment

```
$ kubectl --kubeconfig=cluster/auth/kubeconfig create -f cluster/manifests/kube-scheduler.yaml 
error: error validating "cluster/manifests/kube-scheduler.yaml": error validating data: unexpected type: types.UnixUserID; if you choose to ignore these errors, turn validation off with --validate=false
```

@mengqiy sent a fix https://github.com/kubernetes/kubernetes/pull/47578, but this looks like it's client side only.

Regarding compatibility do we expect kubectl to break like this between versions? It'd be really surprising if, for example, I had a CI process that used kubectl to deployment, then had that break because I upgraded the server.

cc @jamiehannaford

**Comment by liggitt** [comment]
> Regarding compatibility do we expect kubectl to break like this between versions?

No. The core issue is that kubectl is trying to do client-side validation using dynamic discovery of schemas, and is intolerant of unknown schema types. As long as we want to claim +/- 1 version skew support for kubectl, it has to tolerate unknown types in the API and in the schema.

@kubernetes/sig-cli-bugs how do you want to handle this?

**Comment by pwittrock** [comment]
@apelisse Has been working on migrating the validation code to use openapi.  This won't help for 1.7, but maybe he has an idea on how to work around.

To confirm --validate=false is a workaround for kubectl, but doesn't work for cli's that vendor kubectl.

**Comment by liggitt** [comment]
> To confirm --validate=false is a workaround for kubectl, but doesn't work for cli's that vendor kubectl.

There are enough known issues with `--validate` that I'd be concerned about a component that didn't allow opting out

**Comment by foxish** [comment]
Also affecting helm, https://github.com/kubernetes/charts/issues/1294

**Comment by pwittrock** [comment]
Can help just enable opting out of validation for some charts?

**Comment by pwittrock** [comment]
> There are enough known issues with --validate that I'd be concerned about a component that didn't allow opting out

FYI, my comment was meant as a question - does --validate=false workaround the issue for kubectl?

**Comment by smarterclayton** [comment]
Putting this into the milestone until implications are known

**Comment by pwittrock** [comment]
FWIW, any fix we make in 1.7 will not resolve the issue for helm versions vendoring in kubectl 1.6 libraries.  Helm would need to update the vendored kubectl libs to 1.7, which would resolve the issue without any code changes in kubectl anyway.

Kubectl not properly validating fields with types it doesn't have compiled is unfortunate, but IIRC this has been around for several releases, and we have managed to live with it.  We are in the process of rewriting the validation, but it won't be ready for 1.7.

**Comment by pwittrock** [comment]
@smarterclayton I am going to move this out of the milestone, since the issue is with using 1.6 client code on 1.7 objects.  AFAIK, this issue described can't be fixed with code changes to 1.7.

**Comment by pwittrock** [comment]
Will address this with the validation logic in 1.8

**Comment by pwittrock** [comment]
> @mengqiy sent a fix #47578, but this looks like it's client side only.

Great.  I don't think there is a server side fix for this since the validate failures occurs on the client side.

@ericchiang 

> Regarding compatibility do we expect kubectl to break like this between versions? It'd be really surprising if, for example, I had a CI process that used kubectl to deploy, then had that break because I upgraded the server.

No, kubectl is intended to work across skewed client / server versions.  There are many known issues with this though since many kubectl commands use compiled in types for one thing or another.

FWIW, the break wouldn't occur because you upgraded the server, it would happen when you introduced newly introduced fields that weren't known to the client.

**Comment by ericchiang** [comment]
> FWIW, the break wouldn't occur because you upgraded the server, it would happen when you introduced newly introduced fields that weren't known to the client.

Just to be clear, the fields existed in 1.6, the swagger validation changed in 1.7. So the same manifest that can be created against a 1.6 API server breaks against a 1.7 API server.

**Comment by mengqiy** [comment]
> I don't think there is a server side fix for this since the validate failures occurs on the client side.

The change in https://github.com/kubernetes/kubernetes/pull/44714 is an unexpected breaking change for kubectl. A server-side "fix" may be reverting the change of `*int64` -> `*UnixUserID`.

**Comment by pwittrock** [comment]
thanks @ericchiang I didn't realize that.  this seems like a breaking change then.

@thockin @smarterclayton what do you want to do regarding #44714 ?  is there anyway to prevent the break besides a rollback?  We could cherry pick the fix into 1.6, but it would still be broken for kubectl clients that are unpatched

**Comment by smarterclayton** [comment]
Either a revert or a targeted fix to the swagger and API generation to
specifically mark this type as unchanged is all we can do I think.

We need to take client validation out of the critical path of any such
config loop projects, and indicate that if you do use it, it's not
guaranteed to work with  version skew (we reserve the right to change the
swagger definition, and the client is failing safe rather than
continuing).  The fact that helm does not let you bypass client validation
is a bug - client validation is doing what we expect it to do: only make
progress if you see what you expect.

**Comment by dchen1107** [comment]
cc/ @kubernetes/kubernetes-release-managers

**Comment by lavalamp** [comment]
I think there is already a spot to enter exceptions to make the OpenAPI spec show correctly.

**Comment by lavalamp** [comment]
For the record, this is not a kubectl bug, it is a bug in the OpenAPI specification that we publish.

**Comment by lavalamp** [comment]
I'll make a quick patch. Mehdy can double-check and/or change it tomorrow when he is in.

**Comment by pwittrock** [comment]
Great, thanks Daniel.  Removing sig/cli

**Comment by lavalamp** [comment]
...I have a patch but the relevant files don't seem to be getting updated. Not sure what is going on.

**Comment by mbohlool** [comment]
Reading this on my phone. Sorry if I missed a point here.

Kubectl does not use OpenApi spec for validation, right? If true, then this is not OpenAPI issue but the old swagger issue.

**Comment by apelisse** [comment]
> Kubectl does not use OpenApi spec for validation, right? If true, then this is not OpenAPI issue but the old swagger issue.

Correct on both counts.

**Comment by mbohlool** [comment]
I think we have three options:

1. Fix go-restful (the library we use to generate old swagger 1.2) to understand type aliases.
2. Define type tag on all uses of UnixGroupID and UnixUserID (I am exploring this option now).
3. Revert #44714

**Comment by mbohlool** [comment]
So solution 2 is possible but it basically revert all swagger 1.2 changes that #44714 made. As the PR did not affect OpenAPI spec, should we just revert #44714 and remove `types.UnixUserID` and `types.UnixGroupID`? I don't see any real value in keeping them if they are not reflected in our swagger and OpenAPI spec.

**Comment by jamiehannaford** [comment]
@mbohlool Isn't there still value in using real types across the wider K8s codebase instead of arbitrary ints? If so (which is what's outlined in the initial issue), I think #2 is better

**Comment by mbohlool** [comment]
Look like the fix is not possible without changing go-restful (either by adding type aliases there or adding support so we can proceed with solution 2). Given the urgency of this, I would say we go with solution 3, reverting #44714.

**Comment by dchen1107** [comment]
I am intent to go with solution 3 here. @lavalamp @smarterclayton @bgrant0607 WDYT?

**Comment by bgrant0607** [comment]
@liggitt @dchen1107 

This broke the OpenAPI spec, not just client-side validation. `types.UnixUserID` is not a valid type. We should not implement option 2. We need to revert this until type aliases are handled correctly.

@mbohlool Please add a test that no types.Foo types are introduced. 

For that matter, `uid` is a string in the v1 API, which is correct, but is `types.UID` in every other API group, which seems incorrect.

**Comment by bgrant0607** [comment]
Ok, sorry. This broke the Swagger 1.2 spec, which hasn't been officially deprecated yet. We should deprecate it in 1.7, but shouldn't break it.

**Comment by mbohlool** [comment]
we don't have any types.Foo in OpenAPI spec today. The OpenAPI spec generator resolve type aliases. I can add another issue for restricting types.Foo to be added to the spec though I need a little more information on that.

**Comment by dchen1107** [comment]
Ok, looks like we reached the lazy consensus on this: reverting https://github.com/kubernetes/kubernetes/pull/44714 for 1.7. 

@mbohlool is going to send the revert pr. Thanks!

**Comment by dchen1107** [comment]
cc/ @kubernetes/kubernetes-release-managers

**Comment by mbohlool** [comment]
yes, but reverting #44714 is not an easy thing. I am trying different methods. Do you guys have any suggestion?

**Comment by liggitt** [comment]
> types.UnixUserID is not a valid type. We should not implement option 2. We need to revert this until type aliases are handled correctly.

I want to understand why it is not considered a valid type. Is it invalid because the field type changed *at all* in swagger, or because it used a custom type as a leaf field that generated clients couldn't deal with?

Since we've settled on reverting this (and this particular type change didn't really accomplish anything externally) it's a moot point here, but there are other API evolutions I anticipate requiring changes to go struct types, which would be reflected in openapi spec changes.

For example, PV and Pod volume sources use the same type today. There is a need to make object references from persistent volume sources (secrets, endpoints, etc) fully-qualified to include both namespace and name. Rather than adding a field to the shared type that would be disallowed in pods, it would make sense to have a distinct volume source type for use by PVs that included the namespace field. Obviously, json and protobuf compatibility would be required, but a go struct type change would be visible in the openapi schema.

**Comment by mbohlool** [comment]
> Obviously, json and protobuf compatibility would be required, but a go struct type change would be visible in the openapi schema.

It is possible to customize OpenAPI to keep that backward compatible too.

**Comment by liggitt** [comment]
> It is possible to customize OpenAPI to keep that backward compatible too.

What do you mean by backwards compatible? The goal would be for the volume source in a PV to gain namespace fields, and the volume source in a pod to remain the same. This means the volume source type referenced by a PV would have to be different than a Pod.

**Comment by mbohlool** [comment]
I think that is another issue we can discuss separately (would be nice if you can add another issue and point to the exact types and proposal of change so we can discuss it). There are tools to check backward compatibility of OpenAPI and I will add a verify script to run one of them to make sure we keep the spec backward compatible. If we are going to do a backward incompatible change on v1 or any stable API version, we need to have really good justification.

**Comment by liggitt** [comment]
> would be nice if you can add another issue and point to the exact types and proposal of change so we can discuss it

sure, relevant issues are
https://github.com/kubernetes/kubernetes/issues/47427
https://github.com/kubernetes/kubernetes/issues/47452

PR adding namespace field and splitting PV and Pod volume source types is https://github.com/kubernetes/kubernetes/pull/47660

can move the conversation about openapi compatibility there

**Comment by bgrant0607** [comment]
@liggitt The reason was this one:
>> because it used a custom type as a leaf field that clients couldn't deal with

For structures, in general, structural equivalence should be good enough. There are some "well known" types that we may want to have stronger guarantees around, but we should probably do that with properties rather than rely on names.

**Comment by bgrant0607** [comment]
Please someone document that policy. :-)

**Comment by paulwalker** [comment]
Using master branch of helm and tiller does not seem to resolve the issue for at least the concourse chart. How can I workaround this for a 1.7.0 GKE cluster?

**Comment by liggitt** [comment]
the API change was reverted for 1.7.0. What error are you seeing against 1.7.0?
